{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ORF finding\n",
    "\n",
    "## Deliverables:\n",
    " - Lab5 notebook with complete inspection cells ( Introduction and Results )\n",
    " - findORFs.py - 50 total points\n",
    " - imports\n",
    "     - sequenceAnalysis (for FastaReader at least )\n",
    "     - your OrfFinder class ( placed in sequenceAnalysis )\n",
    "     \n",
    "## required output files:\n",
    " - tass2ORFdata-ATG-100.txt \n",
    " \n",
    "## classes required (minimally):\n",
    " - OrfFinder\n",
    " - FastAreader\n",
    " \n",
    "## possible extra credit options - 10 additional points\n",
    " - minOrf= [integer]  This allows exclusion of genes with lengths < minOrf  (Note this length includes both start and stop codons)\n",
    " - biggestOrfOnly True | False  This allows reporting of all genes or only the largest within an ORF\n",
    " - startCodons= ['ATG' | 'TTG' | 'GTG']  This parameter produces a list of start codons used in determining genes ( you might convert this to a more useful container like a set.)\n",
    " - stopCodons = ['TAA' | 'TGA' | 'TAG']  This parameter specifies which codons are to be treated as valid stop codons\n",
    " \n",
    "## Due: Monday February 14, 2022 11:59pm\n",
    "\n",
    "# Overview\n",
    "This lab will give you practice with a real algorithm and with formatted output. We will be finding genes in a genome.\n",
    "\n",
    "An open reading frame (ORF) is a region of DNA that is free from stop codons (TAG, TAA, TGA). We can then define a putative gene in that ORF by the presence of a start codon at the 5' end of the ORF and a stop codon at the 3' end. Start codons in some bacteria, archaea and viruses can be ATG, GTG or CTG, and in some cases TTG.  \n",
    "\n",
    "Since genes are made up of codons of three symbols, any particular sequence of DNA can be interpreted in each of 3 overlapping coding frames, and since cellular genomic DNA is double stranded, we can have genes located on both strands, yielding 6 possible coding frames for the many genes in a genome.  Genes can also overlap each other.\n",
    "\n",
    "To limit the number of possible genes that we might find, two basic strategies are sometimes used:\n",
    "\n",
    " - look for genes that are longer than some predefined number (100 bases for example) and\n",
    " - only consider the largest coding region within an ORF, rather than considering all substrings in the ORF that begin with intermediate start codons.\n",
    " \n",
    "A well known ORF finder is available at NCBI - ORF Finder https://www.ncbi.nlm.nih.gov/orffinder/\n",
    "\n",
    "In this assignment, you will write an equivalent tool that provides the same function, though without the graphical display.\n",
    "\n",
    "Use your sequenceAnalysis module in designing a solution. \n",
    "\n",
    "Your design is essential for success.  Spend time considering how best to solve this problem.  Write out pseudocode that describes the core algorithm.  This pseudocode is required to be delivered in comments at the top of your code submission.  Please start early on this assignment, and consider the algorithm carefully.  \n",
    " - Start and stop codons must be in the same frame, \n",
    " - many possible Met codons (or other possible starts) may exist in an ORF, and \n",
    " - genes can have significant overlap.\n",
    " - In some cases, 2 or more ORFS can exist on opposite strands in the same coordinate space.\n",
    "\n",
    "A partial solution exists in the text using regular expressions (Model p 271).  This solution does not handle any of the extra-credit possibilities, and does not handle the required boundary conditions involving gene fragments. Feel free to look, though I did not find this approach helpful. It also uses regular expressions, which are both difficult to read and disallowed for this assignment.\n",
    "\n",
    "Your implementation should be designed to add capability to your sequenceAnalysis module. Please carefully consider where that capability belongs. You are required to make use of your sequenceAnalysis module for most of the code and for FastAreader. You might consider adding  a function to calculate reverse complement. You should also add an OrfFinder class to sequenceAnalysis. For this assignment, you are the designer of your toolbox - choose wisely. You might use this again in your final project.\n",
    "\n",
    "# ORF finder\n",
    "Create a program to analyze a FASTA-formatted file containing a sequence of DNA and find the ORFs (start and stop codons). Your program is called findORFs.py and it:\n",
    "\n",
    " - reads in FASTA-formatted sequence data from a file using STDIN\n",
    " - finds ORFs that are >= 100 nucleotides in length\n",
    " - writes formatted output data to a text file using STDOUT\n",
    " - if there are multiple FASTA-formatted sequences in the file, you need to perform Orf-finding on each of them independently.\n",
    "\n",
    "# Design specification\n",
    "\n",
    "## Input file\n",
    "Input data must make use of STDIN. This will allow commandLine execution using: \n",
    "- findOrfs.py < someInput.\n",
    "I will provide 2 test files for your use.  For debugging purposes, use lab5test.fa, which is provided in Canvas::files.  As your final test, use the tass2.fa genome. I have provided the output generated by my program for your comparison.\n",
    "\n",
    "## Output file\n",
    "\n",
    "Write your tass2 output to a file named: tass2ORFdata-ATG-100.txt using STDOUT.  This file should be generated with the following program features:\n",
    "\n",
    " - The minimum gene size to report is 100 nucleotides (including stop and start)\n",
    " - The only start codon to consider is ATG\n",
    " - The stop codons in DNA form are: TAA TGA TAG\n",
    " - Only the largest putative gene in an ORF is reported\n",
    " - The input file is tass2.fa\n",
    " \n",
    "I have provided the output generated using a minimum gene size of 300, which you can use for testing. \n",
    "\n",
    "## Gene finding\n",
    "\n",
    "Your  program should find genes in each of the six possible coding frames.  The genes that you find will begin with a start codon and end with a stop codon, with the exception of those gene-fragments that might be located on the ends of the sequence (see boundary conditions).\n",
    "\n",
    "## Boundary conditions\n",
    "\n",
    "You may find that an ORF exists at either or both termini of the sequence you are examining.  This may mean that you will only be seeing a gene fragment located at the sequence termini. \n",
    "\n",
    "For example, consider the sequence:<br>\n",
    "AAA AAA AAA TGA CCC CCC ...\n",
    "\n",
    "Here, we see a stop codon with a set of AAA codons located upstream.  In this case, assume that a start codon exists upstream of our sequence, and report this as a gene in Frame 1, starting at position 1, ending at position 12 and having a length of 12.  \n",
    "\n",
    "The same may occur on the 3' end of the sequence, for example: <br>\n",
    "A AAA AAA ATG CCC CCC CCC CC  \n",
    "\n",
    "In this case, report a gene in Frame 2 starting at position 8, ending at position 21 and having a length of 14.  \n",
    "\n",
    "And, of course genes can exist on the opposite strand, so: <br>\n",
    "TTA AAA AAA AAA CAT CC<br>\n",
    "would be reported in Frame -3, starting in position 1, ending at position 15, with length 15.\n",
    "\n",
    "NCBI ORF finder is a good test case to see what your output should look like.  THe NCBI program does not handle this boundary case however.\n",
    "\n",
    "## Sort Order\n",
    "\n",
    "Your output should be sorted by decreasing ORF size, and in cases where multiple ORFs have the same size, sort those by the left position of the gene.  (build a key that makes this easy).\n",
    "\n",
    "## Minimum gene size (option)\n",
    "\n",
    "The minimum gene size to report is 100 nucleotides.  This count of nucleotides includes the start codon and stop codon.\n",
    "\n",
    "## Largest ORF (option)\n",
    "\n",
    "In any ORF, there may be multiple start codons present.  For the default assignment, only consider the largest gene in an ORF. For the extra-credit assignment, if this (largest ORF) option is not set, report all ORFs found (subject to the minimum size option).\n",
    "\n",
    "## Start codon(s) (option)\n",
    "\n",
    "For the default part of the program, only consider ATG as a start codon. For extra credit, use an input parameter that specifies a set of start codons.\n",
    "\n",
    "## Stop codon(s) (option)\n",
    "\n",
    "For the default part of the program, only consider TAG, TAA and TGA as stop codons. For extra credit,  use an input parameter that specifies the set of stop codons.\n",
    "\n",
    "## Report output\n",
    "\n",
    "All positions reported use the coordinate position of the top strand, and the first position in the sequence (on the left) is position 1.  This coordinate system is used for all 6 coding frames.\n",
    "\n",
    "Frames 1 through 3 can be determined using the start position of the gene.  Assuming the position is p, (p mod 3) + 1 will provide the coding frame.  For genes on the bottom strand, the frame is based on the right hand side of the sequence (see below).\n",
    "\n",
    "For each putative gene, your report must provide:\n",
    "\n",
    " - coding frame. for the top strand, a gene starting in the first position in the sequence is in frame 1.  For the bottom strand, frames are numbered based on the right-hand coordinate position of the sequence.  So, if the sequence is 100 bases long, and you have a gene in positions 89..100, that gene is in frame -1. You can still use the coordinate positions, but the calculation is:  - ( (len(sequence) - last position) mod 3 + 1).  Remember that the last position in this case corresponds to the start codon of the gene on the bottom strand.  These conventions make sense if you think about the reverse complement of the original sequence. The coding frame should be reported with a + or - as needed to describe the coding frame\n",
    " - start position. This position should correspond to the first base of the start codon (top strand) or last base of the stop codon (bottom strand), or the terminus of the genome if this is a gene fragment.\n",
    " - end position. This position should correspond to the first base of the start codon (bottom strand) or last base of the stop codon (top strand), or the terminus of the genome if this is a gene fragment.\n",
    " - length. This includes the start and end codon\n",
    " \n",
    "\n",
    "Line 1 - the header line from the original sequence (without >)\n",
    "\n",
    "Lines 2 .... (use '{:+d} {:>5d}..{:>5d} {:>5d}' or equivalent)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+1 57166..61908  4743 <br>\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra Credit - 10 points possible\n",
    " - Allow your program to optionally use a set of start codons - ATG, GTG, TTG for example.\n",
    " - Allow your program to optionally use a set of stop codons - TAG, TGA, TAA for example.\n",
    " - Allow your program to print every putative gene in an ORF instead of only the largest.\n",
    " - Allow your program to have varying ORF size minimums \n",
    "\n",
    "# Submit your code and answers\n",
    "For this lab, you should upload the following three files as attachments:\n",
    "\n",
    " - your sequenceAnalysis.py module\n",
    " - findORFs.py, include design in # comments at the beginning of findORFs.py\n",
    " - tass2ORFdata-ATG-100.txt\n",
    "\n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Include an overview about what your program is designed to do with expected inputs and outputs\n",
    " - Include design (place at the top of your file below your name, group, and description). Include any assumptions or design decisions you made in writing your code\n",
    " - Adhere to the Report format specification\n",
    " - Include docstrings for Program, Module, Class, Methods and Functions\n",
    " - Contain in-line comments\n",
    " \n",
    " Congratulations, you finished your fifth lab assignment!\n",
    "\n",
    "# Hints:\n",
    "## One strand at a time\n",
    "\n",
    "One design solution considers only the top strand to generate gene candidates. We then generate the reverse complement and a second list of gene candidates.  When done in this way, remember to fix the start/end coordinates that you get from the reverse complement solution, since those will be based on the other end of the sequence. \n",
    "\n",
    "An alternative design exists for the reverse strand that does not require generation of the reverse complement. This solution is interesting and may be simpler to consider. This solution scans the bottom strand left to right, finding stop codons first (reverse complement), then looking for start codons (reverse complement).\n",
    "\n",
    "A third solution for the bottom strand would scan right to left, finding starts until a stop is encountered.\n",
    "\n",
    "## Start and Stop handling\n",
    "\n",
    "If you are going to do the extra-credit, then you will need to remember where the starts are until you find a stop.  This is straightforward if you place those start positions in a list, organized by frame. Even if you are not doing the extra-credit, placing starts on a list organized by frame will really simplify your code. This solution would require three lists for each of the three reading frames.\n",
    "\n",
    "As an alternative, you could scan each of the three reading frames seperately, counting by threes in each case. This seems to be the most straightforward way to do this assignment.\n",
    "\n",
    "When you do find a stop, you then consider the start(s) that you found along the way in this ORF.  Notice that the \"longest\" ORF is then defined by the start that is at the beginning of your start list.  Don't forget to clean up this list after you find genes using this stop since it would terminate all of those \"pending\" genes. What would happen if you didn't clean up your start list - you would be finding ORF candidates that had a stop in the middle of them, which means they are not __OPEN__ reading frames.\n",
    "\n",
    "## Termini and gene fragments\n",
    "\n",
    "This is a bit easier if you think that position 0(1 externally) will allways be a start position for a gene in every frame. This is the case where a start codon exists somewhere upstream of our sequence. It would code for a gene that would terminate in our sequence somewhere. This is the classic example of a dangling-stop. We would report this gene fragment beginning at position 0(1) and ending at the end of that in-frame stop. In frame 1, there might actually be a start codon at this position so dont count it twice. In Frames 2 or 3, there are extra bases ahead of our first full codon, so we will always have the possibility that the actual start is upstream of this dna fragment, so.. we report this as a dangling stop case starting at position 0(1).  \n",
    "\n",
    "Dangling starts always end at the end of sequence, even if that position is out of frame. In cases where no start or stop codon is found anywhere in the frame, the entire sequence is then considered as a gene.\n",
    "\n",
    "# Data and output files\n",
    "\n",
    "The cell below is set up for the non-extra credit version of the program with a fixed filename. Your much more usable program will use STDIN and STDOUT in command line fashion. To do this, simplify main so your fastaReader object is reading from STDIN.\n",
    "\n",
    "All of the input files and programs should be in the same folder so you won’t have to specify a path, if your data file is in some other folder, you can provide that like this:<br> ~/Desktop/bme160/someOtherPlace/labData.fa.\n",
    "\n",
    "# Comments\n",
    "\n",
    "To get full credit for this lab, you must turn in your design for each program you submit. Remember that any item you write or include must have proper docstrings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GENE FINDER INTRODUCTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The gene finder algorithm is a well established algorithm offered by the NCBI where we are asked to look through 6 different frames of codons and establish whether a gene is a valid gene based on some criterion. We built a version of this gene finder but only the backend of the code. Still when we were discussing this algorithm in class I thought it made very intuitive sense and infact I made a very similar program in my CSE 30 class where we would find the potential proteins of a DNA sequence finding the start codons and calling it a potential protein once we found one of the End Codons. We then translated the codons into their amino acid and printed out to user output. Therefore I had alot of experience working with frames and it made a lot of sense to me. In terms of algorithm development I used a flag system to check when we found start codons and set flags to keep track of progress. However the user output was something else I had to keep track of and I was surprised to have to keep track of many many lists of start index, end index, frame, and length. Therefore the user output was a lot to keep track of. However not impossible. I really enjoyed this assignment though"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findORFs.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [options] -option1[default] <input >output\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /Users/simonlee/Library/Jupyter/runtime/kernel-64f33db0-86c3-4116-b273-484717b10ec5.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/simonlee/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3445: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Simon Lee (siaulee)\n",
    "# Group Members: Campbell Strand, Mateo Etcheveste, Jonathan Zau\n",
    "\n",
    "from sequenceAnalysis import OrfFinder, FastAreader\n",
    "\n",
    "\"\"\"\n",
    "ORF Finder\n",
    "----------\n",
    "iterate through three frames from 3' - 5'\n",
    "    iterate through codons through splicing i, i+3\n",
    "        if a start codon is found\n",
    "            add its index\n",
    "        if a stop codon is found\n",
    "            find its index\n",
    "            find its length\n",
    "            if length is > minlength\n",
    "                save the gene from orf\n",
    "            reset variables\n",
    "        \n",
    "        check for frame with with no start codon but a end codon\n",
    "            repeat steps from first set of if statements\n",
    "        \n",
    "        check for frame with start codon but no end codon\n",
    "            repeat steps from above\n",
    "\n",
    "**repeat for reverse from 5' to 3' end**\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "class CommandLine() :\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if command is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then command.args.requiredbool will\n",
    "    name that option.\n",
    " \n",
    "    '''\n",
    "    \n",
    "    def __init__(self, inOpts=None) :\n",
    "        '''\n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "        \n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(description = 'Program prolog - a brief description of what this thing does', \n",
    "                                             epilog = 'Program epilog - some other stuff you feel compelled to say', \n",
    "                                             add_help = True, #default is True \n",
    "                                             prefix_chars = '-', \n",
    "                                             usage = '%(prog)s [options] -option1[default] <input >output'\n",
    "                                             )\n",
    "        self.parser.add_argument('-lG', '--longestGene', action = 'store', nargs='?', const=True, default=False, help='longest Gene in an ORF')\n",
    "        self.parser.add_argument('-mG', '--minGene', type=int, choices= (100,200,300,500,1000), default=100, action = 'store', help='minimum Gene length')\n",
    "        self.parser.add_argument('-s', '--start', action = 'append', default = ['ATG'],nargs='?', \n",
    "                                 help='start Codon') #allows multiple list options\n",
    "        self.parser.add_argument('-t', '--stop', action = 'append', default = ['TAG','TGA','TAA'],nargs='?', help='stop Codon') #allows multiple list options\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')  \n",
    "        if inOpts is None :\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "\n",
    "def main():\n",
    "    '''\n",
    "    A program that reads in a fasta file from command line argument and prints out genes from longest to shortest\n",
    "    '''\n",
    "    command = CommandLine()\n",
    "    if command.args.longestGene:\n",
    "        fastaFile = FastAreader()\n",
    "        \n",
    "        # use same file reader code from last week\n",
    "        for header, sequence in fastaFile.readFasta():\n",
    "            print(header)\n",
    "            # create an OrfFinder object\n",
    "            orf_info = OrfFinder(sequence)\n",
    "            \n",
    "            #append list of genes found with minGene argument\n",
    "            orf_info.Orf_search(command.args.minGene)\n",
    "            orf_info.rev_Orf_search(command.args.minGene)\n",
    "            # print statement that sorts our genes by longest length to shortest length\n",
    "            orf_list = sorted(orf_info.found_orfs, key=lambda orf: orf[3], reverse = True)\n",
    "            for frame, start_index, stop_index, seq_length in orf_list:\n",
    "                if seq_length > command.args.minGene: \n",
    "                    print('{:+d} {:>5d}..{:>5d} {:>5d}'.format(frame, start_index, stop_index, seq_length))\n",
    "\n",
    "        \n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# sequenceAnalysis.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3 \n",
    "# Name: Simon Lee (siaulee)\n",
    "# Group Members: Campbell Strand, Mateo Etcheveste, Jonathan Zau\n",
    "\n",
    "'''\n",
    "A module that performs a lot of counts for DNA, codons, and Amino acids\n",
    "'''\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "class ProteinParam :\n",
    "    '''\n",
    "    This class has various methods that calculate a wide range of statistics.\n",
    "    \n",
    "    In addition we utilize some fixed values in the form of a dictionary to assist in computation\n",
    "    '''\n",
    "# These tables are for calculating:\n",
    "#     molecular weight (aa2mw), along with the mol. weight of H2O (mwH2O)\n",
    "#     absorbance at 280 nm (aa2abs280)\n",
    "#     pKa of positively charged Amino Acids (aa2chargePos)\n",
    "#     pKa of negatively charged Amino acids (aa2chargeNeg)\n",
    "#     and the constants aaNterm and aaCterm for pKa of the respective termini\n",
    "#  Feel free to move these to appropriate methods as you like\n",
    "\n",
    "# As written, these are accessed as class attributes, for example:\n",
    "# ProteinParam.aa2mw['A'] or ProteinParam.mwH2O\n",
    "\n",
    "    aa2mw = {\n",
    "        'A': 89.093,  'G': 75.067,  'M': 149.211, 'S': 105.093, 'C': 121.158,\n",
    "        'H': 155.155, 'N': 132.118, 'T': 119.119, 'D': 133.103, 'I': 131.173,\n",
    "        'P': 115.131, 'V': 117.146, 'E': 147.129, 'K': 146.188, 'Q': 146.145,\n",
    "        'W': 204.225,  'F': 165.189, 'L': 131.173, 'R': 174.201, 'Y': 181.189\n",
    "        }\n",
    "\n",
    "    mwH2O = 18.015\n",
    "    aa2abs280= {'Y':1490, 'W': 5500, 'C': 125}\n",
    "\n",
    "    aa2chargePos = {'K': 10.5, 'R':12.4, 'H':6}\n",
    "    aa2chargeNeg = {'D': 3.86, 'E': 4.25, 'C': 8.33, 'Y': 10}\n",
    "    aaNterm = 9.69\n",
    "    aaCterm = 2.34\n",
    "\n",
    "    def __init__ (self, protein):\n",
    "        '''\n",
    "        Constructor method for the ProteinParam class\n",
    "        '''\n",
    "        protein = protein.upper() # converts every sequence to uppercase - useful across all programs\n",
    "        splitSeq = ''.join(protein).split() # split at every space\n",
    "        self.proteinSeq = ''.join(splitSeq).upper() # conncatenate strings\n",
    "        \n",
    "        # constructs dictionary that counts \n",
    "        self.aminoDict = {key:self.proteinSeq.count(key) \n",
    "                       for key in ProteinParam.aa2mw.keys()}\n",
    "        \n",
    "\n",
    "    def aaCount (self):\n",
    "        '''\n",
    "        takes in no arguments and counts the occurence of amino acids\n",
    "        '''\n",
    "        aaCount = 0\n",
    "        for aa in self.proteinSeq: # goes through user input sequence\n",
    "            if aa in self.aa2mw.keys(): # checks to see if amino acid dict keys match seq amino acids\n",
    "                aaCount += 1\n",
    "        return aaCount  \n",
    "\n",
    "    def pI (self):\n",
    "        '''\n",
    "        takes in no arguments. theoretical isolelectric point can be estimated \n",
    "        by finding the particular pH that yields a neutral net Charge (closest to 0)\n",
    "        '''\n",
    "        charge = []\n",
    "        rangeList = np.arange(0.0,14.01,0.01) # makes numpy array from 0 to 14 spaced by 0.01\n",
    "        for pH in rangeList:\n",
    "            charge.append(self._charge_(pH))\n",
    "        value = min(charge, key=abs) # finds smallest value - straight from my codelab \n",
    "        index = charge.index(value) * 0.01 # finds index which correlates to pH \n",
    "        return index\n",
    "\n",
    "    def aminoDictosition (self) :\n",
    "        '''\n",
    "        takes in no arguments. returns a dictionary of amino acid counts from constructor\n",
    "        '''   \n",
    "        return self.aminoDict \n",
    "\n",
    "    def _charge_ (self,pH):\n",
    "        '''\n",
    "        takes in pH as a parameter to find net charge at certain pH level. \n",
    "        This method calculates the net charge on the protein at a specific pH\n",
    "        '''\n",
    "        pos, neg = 0.0, 0.0\n",
    "        for aa in self.proteinSeq:   \n",
    "            # summation for positive charge\n",
    "            if aa in self.aa2chargePos.keys():   \n",
    "                pos += (10 ** self.aa2chargePos[aa])/ (10 ** self.aa2chargePos[aa] + (10 ** pH))  \n",
    "            \n",
    "            # summation for negative charge\n",
    "            elif aa in self.aa2chargeNeg.keys():\n",
    "                neg += (10**pH)/(10**self.aa2chargeNeg[aa]+ (10 ** pH))\n",
    "                \n",
    "            else:\n",
    "                continue\n",
    "        # adds the N and C terminus\n",
    "        pos += (10**self.aaNterm)/(10**self.aaNterm + 10**pH)\n",
    "        neg += (10**pH)/(10**self.aaCterm+10**pH)\n",
    "        netCharge = pos - neg\n",
    "        \n",
    "        return netCharge\n",
    "\n",
    "    def molarExtinction (self):\n",
    "        '''\n",
    "        takes in no arguments. The extinction coefficient indicates how much light \n",
    "        a protein absorbs at a certain wavelength. From the molar extinction \n",
    "        coefficient of tyrosine, tryptophan and cystine at a given wavelength, \n",
    "        the extinction coefficient of the native protein in water can be computed:\n",
    "        \n",
    "        𝐸=(𝑁_𝑌*𝐸_𝑌)+(𝑁_𝑊*𝐸_𝑊)+(𝑁_𝐶*𝐸_𝐶)\n",
    "        '''\n",
    "        # equation above\n",
    "        molarExt = float((self.aminoDict.get('Y')*self.aa2abs280['Y']) + (self.aminoDict.get('W')*self.aa2abs280['W']) + (self.aminoDict.get('C')*self.aa2abs280['C']))\n",
    "        return molarExt\n",
    "                         \n",
    "    def massExtinction (self):\n",
    "        '''\n",
    "        takes in no arguments. Using molar extinction and molecular weight we can calculate massExtinction\n",
    "        '''\n",
    "        myMW =  self.molecularWeight()\n",
    "        return self.molarExtinction() / myMW if myMW else 0.0\n",
    "\n",
    "    def molecularWeight (self):\n",
    "        '''\n",
    "        takes in no arguments. Calculates amino acids molecular weights and excluding the waters \n",
    "        that are released with peptide bond formation\n",
    "        '''\n",
    "        sumVal = 0\n",
    "        water = self.mwH2O * (self.aaCount() - 1) # calculates water\n",
    "        \n",
    "        # runs summation that calculates molecular weight\n",
    "        for aa in self.proteinSeq:\n",
    "            if aa in self.aa2mw.keys():\n",
    "                sumVal += self.aa2mw.get(aa)\n",
    "        \n",
    "        molecWeight = sumVal - water\n",
    "        return molecWeight\n",
    "    \n",
    "    \n",
    "class NucParams:\n",
    "    '''\n",
    "    The NucParams class constructs a multitude of counts using dictionaries\n",
    "    '''\n",
    "    rnaCodonTable = {\n",
    "    # RNA codon table\n",
    "    # U\n",
    "    'UUU': 'F', 'UCU': 'S', 'UAU': 'Y', 'UGU': 'C',  # UxU\n",
    "    'UUC': 'F', 'UCC': 'S', 'UAC': 'Y', 'UGC': 'C',  # UxC\n",
    "    'UUA': 'L', 'UCA': 'S', 'UAA': '-', 'UGA': '-',  # UxA\n",
    "    'UUG': 'L', 'UCG': 'S', 'UAG': '-', 'UGG': 'W',  # UxG\n",
    "    # C\n",
    "    'CUU': 'L', 'CCU': 'P', 'CAU': 'H', 'CGU': 'R',  # CxU\n",
    "    'CUC': 'L', 'CCC': 'P', 'CAC': 'H', 'CGC': 'R',  # CxC\n",
    "    'CUA': 'L', 'CCA': 'P', 'CAA': 'Q', 'CGA': 'R',  # CxA\n",
    "    'CUG': 'L', 'CCG': 'P', 'CAG': 'Q', 'CGG': 'R',  # CxG\n",
    "    # A\n",
    "    'AUU': 'I', 'ACU': 'T', 'AAU': 'N', 'AGU': 'S',  # AxU\n",
    "    'AUC': 'I', 'ACC': 'T', 'AAC': 'N', 'AGC': 'S',  # AxC\n",
    "    'AUA': 'I', 'ACA': 'T', 'AAA': 'K', 'AGA': 'R',  # AxA\n",
    "    'AUG': 'M', 'ACG': 'T', 'AAG': 'K', 'AGG': 'R',  # AxG\n",
    "    # G\n",
    "    'GUU': 'V', 'GCU': 'A', 'GAU': 'D', 'GGU': 'G',  # GxU\n",
    "    'GUC': 'V', 'GCC': 'A', 'GAC': 'D', 'GGC': 'G',  # GxC\n",
    "    'GUA': 'V', 'GCA': 'A', 'GAA': 'E', 'GGA': 'G',  # GxA\n",
    "    'GUG': 'V', 'GCG': 'A', 'GAG': 'E', 'GGG': 'G'  # GxG\n",
    "    }\n",
    "    dnaCodonTable = {key.replace('U','T'):value for key, value in rnaCodonTable.items()}\n",
    "\n",
    "    def __init__ (self, inString=''):\n",
    "        '''\n",
    "        Constructor Method for NucParams program. Constructs all ionaries taking optional user input\n",
    "        '''\n",
    "        # nucleotide bases Dictionary\n",
    "        self.validNuc = ['A','C','G','T','U','N']\n",
    "        self.nucComp = {nucleo: 0 for nucleo in self.validNuc}\n",
    "        \n",
    "        # builds a Dictionary for amino acids\n",
    "        self.aaComp = {amino: 0 for amino in NucParams.rnaCodonTable.values()}\n",
    "        \n",
    "        # builds a Dictionary for codons\n",
    "        self.codonComp = {codon: 0 for codon in NucParams.rnaCodonTable.keys()}\n",
    "        \n",
    "        self.addSequence(inString)\n",
    "        \n",
    "        \n",
    "    def addSequence (self, inSeq):\n",
    "        '''\n",
    "        Method that does all the intangibles of the code\n",
    "        '''\n",
    "        \n",
    "        # cleans up sequence\n",
    "        temp = ''.join(inSeq.upper()).split()\n",
    "        usr = ''.join(temp)\n",
    "        seq = ''\n",
    "        \n",
    "        # gets rid of invalid characters (non - ACTGUN) and counts nucleotide bases\n",
    "        for nucleo in usr:\n",
    "            if nucleo in self.nucComp.keys():\n",
    "                seq += str(nucleo)\n",
    "                self.nucComp[nucleo] += 1\n",
    "        \n",
    "        # transcribes DNA to RNA\n",
    "        seq = seq.replace(\"T\", \"U\")\n",
    "        \n",
    "        # constructs codon list within a sequence \n",
    "        codonsList = []\n",
    "        for i in range(0, len(seq), 3):\n",
    "            codonsList.append(seq[i:i+3])\n",
    "        \n",
    "        # translation from codon to amino acid\n",
    "        for codon in codonsList:\n",
    "            if codon in NucParams.rnaCodonTable.keys():\n",
    "                self.aaComp[NucParams.rnaCodonTable[codon]] += 1\n",
    "                self.codonComp[codon] += 1\n",
    "            \n",
    "    def aaComposition(self):\n",
    "        '''\n",
    "        returns a Dictionary of the amino acid counts\n",
    "        '''\n",
    "        return self.aaComp\n",
    "    def nucComposition(self):\n",
    "        '''\n",
    "        returns a Dictionary of counts of valid nucleotides found in the analysis. (ACGTNU}\n",
    "        '''\n",
    "        return self.nucComp\n",
    "    def codonComposition(self):\n",
    "        '''\n",
    "        returns a Dictionary that counts codons.\n",
    "        '''\n",
    "        return self.codonComp\n",
    "    def nucCount(self):\n",
    "        '''\n",
    "        returns an integer value, summing every valid nucleotide {ACGTUN} \n",
    "        '''\n",
    "        return sum(self.nucComp.values())\n",
    "\n",
    "import sys\n",
    "class FastAreader :\n",
    "    ''' \n",
    "    Define objects to read FastA files.\n",
    "    \n",
    "    instantiation: \n",
    "    thisReader = FastAreader ('testTiny.fa')\n",
    "    usage:\n",
    "    for head, seq in thisReader.readFasta():\n",
    "        print (head,seq)\n",
    "    '''\n",
    "    def __init__ (self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    "            \n",
    "    def doOpen (self):\n",
    "        ''' Handle file opens, allowing STDIN.'''\n",
    "        if self.fname is '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "        \n",
    "    def readFasta (self):\n",
    "        ''' Read an entire FastA record and return the sequence header/sequence'''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with self.doOpen() as fileH:\n",
    "            \n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header,sequence\n",
    "\n",
    "        \n",
    "class OrfFinder():\n",
    "    '''\n",
    "    class that finds genes in open reading frames from FASTA file\n",
    "    '''\n",
    "\n",
    "    def __init__(self, sequence):\n",
    "        '''\n",
    "        Constructor method that sets up all seqence orf algorithm\n",
    "        '''\n",
    "        self.sequence = sequence\n",
    "        self.complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "        self.stopCodons = ['TAG','TAA','TGA']\n",
    "        self.startCodons = ['ATG']\n",
    "        self.found_orfs = [] \n",
    "\n",
    "    def reverse(self):\n",
    "        '''\n",
    "        returns the reverse complement of the DNA strands\n",
    "        '''\n",
    "        rev = list(self.sequence)\n",
    "        rev = reversed([self.complement.get(nuc, nuc) for nuc in rev])\n",
    "        rev = ''.join(rev)\n",
    "        \n",
    "        reverse_seq = rev\n",
    "        \n",
    "        return reverse_seq\n",
    "\n",
    "    def add_Orf(self, frame, start_index, stop_index, seq_length):\n",
    "        '''\n",
    "        adds genes found to orfs list\n",
    "        '''\n",
    "        self.found_orfs.append([frame, start_index, stop_index, seq_length])\n",
    "        \n",
    "    def Orf_search(self, minlength):\n",
    "        '''\n",
    "        Find genes in Open reading frames on the 3'-5' strand \n",
    "        '''\n",
    "        pos = [] # empty list that tracks the start \n",
    "        startPos = False\n",
    "        codonIterate = False\n",
    "\n",
    "        for frame in range(0,3): \n",
    "            startPos = False  \n",
    "            codonIterate = False\n",
    "            pos = []\n",
    "            for i in range(frame, len(self.sequence), 3):\n",
    "                # reads a codon at a time\n",
    "                codon = self.sequence[i:i+3] \n",
    "                # begins counting sequence length after first codon is found\n",
    "                if codon in self.startCodons:\n",
    "                    startPos = True\n",
    "                    codonIterate = True\n",
    "                    pos.append(int(i)) # updates position\n",
    "                \n",
    "                # when stop codon is found, we check whether it is a valid orf\n",
    "                if (codon in self.stopCodons) and startPos == True:\n",
    "                    start = pos[0] + 1 - frame\n",
    "                    stop = i + 3\n",
    "                    if frame == 0:\n",
    "                        length = stop - start + 1\n",
    "                    if frame == 1:\n",
    "                        length = stop - start\n",
    "                    if frame == 2: \n",
    "                        length = stop - start - 1\n",
    "                    if length > minlength:\n",
    "                        self.add_Orf((frame%3)+1, start, stop, length)\n",
    "                            \n",
    "                    pos = []\n",
    "                    \n",
    "                    startPos = False\n",
    "                    codonIterate = True\n",
    "                \n",
    "                # edge case for no start codon but found stop codon, dangling end\n",
    "                if (codon in self.stopCodons) and (codonIterate == False):  \n",
    "                    start = 1\n",
    "                    stop = i + 3\n",
    "                    length = stop - start + 1\n",
    "                    if length > minlength:\n",
    "                        self.add_Orf((frame%3)+1, start, stop, length)\n",
    "                    pos = []\n",
    "                    \n",
    "                    codonIterate = True\n",
    "                    \n",
    "            # edge case for start codon but found no stop codon, dangling stop\n",
    "            if startPos == True:  \n",
    "                start = pos[0] + 1\n",
    "                stop = len(self.sequence)\n",
    "                length = stop - start + 1\n",
    "                if length > minlength:\n",
    "                    self.add_Orf((frame%3)+1, start, stop, length)\n",
    "        return self.found_orfs\n",
    "\n",
    "    def rev_Orf_search(self, minlength): #go trhough test file and comment out working lines, then compare orfs by hand to find error\n",
    "        '''\n",
    "        Find genes in Open reading frames on the 5'-3' strand \n",
    "        '''\n",
    "        # extract a reverse complemented sequence\n",
    "        revSeq = self.reverse()\n",
    "        pos = []\n",
    "        startPos = False\n",
    "        codonIterate = False\n",
    "        \n",
    "        # repeat Orf finder algorithm - first iterate through 3 frames\n",
    "        for frame in range(0, 3):  \n",
    "            startPos = False  \n",
    "            codonIterate = False\n",
    "            pos = []\n",
    "            \n",
    "            # check through each codon\n",
    "            for i in range(frame, len(revSeq), 3):\n",
    "                codon = revSeq[i:i + 3]  \n",
    "                \n",
    "                # if codon is a start codon start tracking index and length\n",
    "                if codon in self.startCodons:  \n",
    "                    pos.append(i)\n",
    "                    startPos = True\n",
    "                    codonIterate = True\n",
    "                \n",
    "                # if codon is a stop codon start calculating  length\n",
    "                if (codon in self.stopCodons) and (startPos == True):  \n",
    "                    stop = len(revSeq) - pos[0]\n",
    "                    start = len(revSeq) - (i+2)\n",
    "                    length = stop - start + 1\n",
    "                    if length > minlength:\n",
    "                        self.add_Orf(((frame%3) + 1)* -1, start, stop, length)\n",
    "                    pos = []\n",
    "                    startPos = False\n",
    "                    codonIterate = True\n",
    "                \n",
    "                # edge case for no start codon but found stop codon, dangling end\n",
    "                if (codon in self.stopCodons) and (codonIterate == False):  \n",
    "                    start = len(revSeq) - i - 1\n",
    "                    stop = len(revSeq)\n",
    "                    length = stop - start + 1\n",
    "                    if length > minlength:\n",
    "                        self.add_Orf(((frame%3) + 1)* -1, start, stop, length)\n",
    "                    pos = []\n",
    "                    codonIterate = True\n",
    "            \n",
    "            # edge case for start codon but found no stop codon, dangling stop\n",
    "            if startPos == True:  \n",
    "                start =  pos[0] + 1\n",
    "                stop = 1\n",
    "                if length > minlength:\n",
    "                    length = stop - start + 1\n",
    "                self.add_Orf((-1 * (frame%3) + 1), start, stop, length)\n",
    "        return self.found_orfs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Command Line Argument input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python3 findOrfs.py  <tass2.fa  >tass2ORFdata-ATG-100.txt -mG 300 -lG -s ATG -s GTG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findORFs output using tass2.fa  -mG=300 -lG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tass2 NODE_159_length_75728_cov_97.549133\n",
    "+1 57166..61908  4743\n",
    "-1  8192..11422  3231\n",
    "+2 65963..69004  3042\n",
    "-3 14589..16862  2274\n",
    "-2  2968.. 4872  1905\n",
    "+1 64093..65952  1860\n",
    "-3    30.. 1694  1665\n",
    "+1 69475..71052  1578\n",
    "+1 48805..50223  1419\n",
    "+1 47398..48798  1401\n",
    "-3 29133..30500  1368\n",
    "-1 40922..42250  1329\n",
    "-2 19975..21270  1296\n",
    "+3 72273..73559  1287\n",
    "-1 24482..25639  1158\n",
    "-1  6689.. 7804  1116\n",
    "+1 50251..51366  1116\n",
    "-3 27501..28601  1101\n",
    "+2 63038..64078  1041\n",
    "+3 62019..63038  1020\n",
    "-2 42271..43263   993\n",
    "+3 51864..52805   942\n",
    "+1 45484..46371   888\n",
    "-3 11433..12317   885\n",
    "+2 74357..75184   828\n",
    "+2 71576..72394   819\n",
    "+3 46341..47138   798\n",
    "-2 18613..19407   795\n",
    "+1 55642..56388   747\n",
    "-1 16940..17632   693\n",
    "-2 13288..13974   687\n",
    "-3 26115..26801   687\n",
    "-1 21338..21994   657\n",
    "-1 30998..31654   657\n",
    "-2 12601..13251   651\n",
    "-2  4894.. 5532   639\n",
    "-3 32592..33221   630\n",
    "-1 39914..40525   612\n",
    "+1 53977..54588   612\n",
    "+3 54588..55193   606\n",
    "-3 33234..33809   576\n",
    "-3 22002..22559   558\n",
    "-3 23859..24413   555\n",
    "-2  1945.. 2490   546\n",
    "+1 73861..74370   510\n",
    "-2  6214.. 6696   483\n",
    "+1 16324..16806   483\n",
    "-1 22556..23032   477\n",
    "+3 32808..33260   453\n",
    "+2 53324..53776   453\n",
    "+1 29056..29502   447\n",
    "-2 36286..36729   444\n",
    "+3 51396..51833   438\n",
    "+2 55196..55627   432\n",
    "-1  2468.. 2896   429\n",
    "-2 31798..32220   423\n",
    "+1 52891..53307   417\n",
    "-2 30595..30996   402\n",
    "-2  5809.. 6204   396\n",
    "+1     1..  393   393\n",
    "-2 35740..36129   390\n",
    "-3 34542..34925   384\n",
    "+2 56474..56857   384\n",
    "-1 17888..18268   381\n",
    "-3 23004..23381   378\n",
    "-3 52968..53336   369\n",
    "-1 32207..32572   366\n",
    "-1 23396..23752   357\n",
    "+2 36419..36775   357\n",
    "-2 18268..18609   342\n",
    "+2 44594..44935   342\n",
    "+1 43714..44049   336\n",
    "+2  6557.. 6886   330\n",
    "-2 34927..35253   327\n",
    "-3 34131..34451   321\n",
    "+1 30778..31095   318\n",
    "+2 24974..25288   315\n",
    "+1  1246.. 1557   312\n",
    "-3 35253..35561   309\n",
    "-2 39160..39468   309\n",
    "-2 68932..69234   303"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findORFs output using lab5test.fa at mG=0 cutoff\n",
    "\n",
    "(Note: I have included the sequence of the gene candidate as an additional column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test\n",
    "+1     1..    9     9 ATGATGTAA\n",
    "+3     1..    9     9 ATGATGTAA\n",
    "-1     1..    9     9 TTACATCAT\n",
    "-2     1..    9     9 TTACATCAT\n",
    "-3     1..    9     9 TTACATCAT\n",
    "+1     4..    9     6 ATGTAA\n",
    "+2     1..    4     4 ATGA\n",
    "test2\n",
    "+1     1..   10    10 AATGATGTAA\n",
    "+2     1..   10    10 AATGATGTAA\n",
    "-1     1..   10    10 TTACATCATT\n",
    "-2     1..   10    10 TTACATCATT\n",
    "-3     1..   10    10 TTACATCATT\n",
    "+2     2..   10     9 ATGATGTAA\n",
    "+2     5..   10     6 ATGTAA\n",
    "+3     1..    5     5 AATGA\n",
    "test3\n",
    "+2     1..   11    11 AAATGATGTAA\n",
    "+3     1..   11    11 AAATGATGTAA\n",
    "-1     1..   11    11 TTACATCATTT\n",
    "-2     1..   11    11 TTACATCATTT\n",
    "-3     1..   11    11 TTACATCATTT\n",
    "+3     3..   11     9 ATGATGTAA\n",
    "+1     1..    6     6 AAATGA\n",
    "+3     6..   11     6 ATGTAA\n",
    "test-1\n",
    "+1     1..    9     9 TTACATCAT\n",
    "+2     1..    9     9 TTACATCAT\n",
    "+3     1..    9     9 TTACATCAT\n",
    "-1     1..    9     9 ATGATGTAA\n",
    "-3     1..    9     9 ATGATGTAA\n",
    "-1     1..    6     6 ATGTAA\n",
    "-2     6..    9     4 ATGA\n",
    "test-2\n",
    "+1     1..   10    10 TTACATCATA\n",
    "+2     1..   10    10 TTACATCATA\n",
    "+3     1..   10    10 TTACATCATA\n",
    "-1     1..   10    10 TATGATGTAA\n",
    "-2     1..   10    10 TATGATGTAA\n",
    "-2     1..    9     9 ATGATGTAA\n",
    "-2     1..    6     6 ATGTAA\n",
    "-3     6..   10     5 TATGA\n",
    "test-3\n",
    "+1     1..   11    11 TTACATCATAA\n",
    "+2     1..   11    11 TTACATCATAA\n",
    "+3     1..   11    11 TTACATCATAA\n",
    "-2     1..   11    11 TTATGATGTAA\n",
    "-3     1..   11    11 TTATGATGTAA\n",
    "-3     1..    9     9 ATGATGTAA\n",
    "-3     1..    6     6 ATGTAA\n",
    "-1     6..   11     6 TTATGA\n",
    "test1A\n",
    "+3     1..   10    10 ATGATGTAAA\n",
    "-1     1..   10    10 TTTACATCAT\n",
    "-2     1..   10    10 TTTACATCAT\n",
    "-3     1..   10    10 TTTACATCAT\n",
    "+1     1..    9     9 ATGATGTAA\n",
    "+1     4..    9     6 ATGTAA\n",
    "+2     1..    4     4 ATGA\n",
    "test2A\n",
    "+1     1..   11    11 AATGATGTAAA\n",
    "-1     1..   11    11 TTTACATCATT\n",
    "-2     1..   11    11 TTTACATCATT\n",
    "-3     1..   11    11 TTTACATCATT\n",
    "+2     1..   10    10 AATGATGTAA\n",
    "+2     2..   10     9 ATGATGTAA\n",
    "+2     5..   10     6 ATGTAA\n",
    "+3     1..    5     5 AATGA\n",
    "test3A\n",
    "+2     1..   12    12 AAATGATGTAAA\n",
    "-1     1..   12    12 TTTACATCATTT\n",
    "-2     1..   12    12 TTTACATCATTT\n",
    "-3     1..   12    12 TTTACATCATTT\n",
    "+3     1..   11    11 AAATGATGTAA\n",
    "+3     3..   11     9 ATGATGTAA\n",
    "+1     1..    6     6 AAATGA\n",
    "+3     6..   11     6 ATGTAA\n",
    "test-1A\n",
    "+1     1..   10    10 ATTACATCAT\n",
    "+2     1..   10    10 ATTACATCAT\n",
    "+3     1..   10    10 ATTACATCAT\n",
    "-3     1..   10    10 ATGATGTAAT\n",
    "-1     2..   10     9 ATGATGTAA\n",
    "-1     2..    7     6 ATGTAA\n",
    "-2     7..   10     4 ATGA\n",
    "test-2A\n",
    "+1     1..   12    12 AATTACATCATA\n",
    "+2     1..   12    12 AATTACATCATA\n",
    "+3     1..   12    12 AATTACATCATA\n",
    "-1     1..   12    12 TATGATGTAATT\n",
    "-2     3..   12    10 TATGATGTAA\n",
    "-2     3..   11     9 ATGATGTAA\n",
    "-2     3..    8     6 ATGTAA\n",
    "-3     8..   12     5 TATGA\n",
    "test-3A\n",
    "+1     1..   13    13 AATTACATCATAA\n",
    "+2     1..   13    13 AATTACATCATAA\n",
    "+3     1..   13    13 AATTACATCATAA\n",
    "-2     1..   13    13 TTATGATGTAATT\n",
    "-3     3..   13    11 TTATGATGTAA\n",
    "-3     3..   11     9 ATGATGTAA\n",
    "-3     3..    8     6 ATGTAA\n",
    "-1     8..   13     6 TTATGA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inspection Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Jonathan (Feb 8) - It was really helpful to see Simon's code as I have had a chance to get some inspiration for my own implementation. Simon's code follows our design that we drew together in class and I thought it was very intuitive and the logic was easy to follow. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cambell (Feb 8) - Simon has been really helpful in this lab. It is always nice to have a reference code in our group and to see him bring his design exactly and use that is really something to see. He has been a really helpful resource throughout the whole class helping everyone in our group."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mateo (Feb 8) - When Simon told us he was done with his code over the weekend I was pretty surprised considering we had to implement a gene finder algorithm. However I know he has a lot of previous experience coding so I am also not too surprised. Good job Simon!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
